// @web3auth/wallet-connect-v2-adapter/dist/lib.cjs/index.js
'use strict';

var config = require('./config.js');
var walletConnectV2adapter = require('./walletConnectV2adapter.js');
var WalletConnectV2Provider = require('./WalletConnectV2Provider.js');



exports.DEFAULT_EIP155_METHODS = config.DEFAULT_EIP155_METHODS;
exports.DEFAULT_EIP_155_EVENTS = config.DEFAULT_EIP_155_EVENTS;
exports.DEFAULT_SOLANA_EVENTS = config.DEFAULT_SOLANA_EVENTS;
exports.DEFAULT_SOLANA_METHODS = config.DEFAULT_SOLANA_METHODS;
exports.SOLANA_CAIP_CHAIN_MAP = config.SOLANA_CAIP_CHAIN_MAP;
exports.getNamespacesFromChains = config.getNamespacesFromChains;
exports.getRequiredNamespaces = config.getRequiredNamespaces;
exports.getSupportedEventsByNamespace = config.getSupportedEventsByNamespace;
exports.getSupportedMethodsByNamespace = config.getSupportedMethodsByNamespace;
exports.getWalletConnectV2Settings = config.getWalletConnectV2Settings;
exports.WalletConnectV2Adapter = walletConnectV2adapter.WalletConnectV2Adapter;
exports.WalletConnectV2Provider = WalletConnectV2Provider.WalletConnectV2Provider;
