// @web3auth/default-evm-adapter/dist/lib.cjs/injectedEvmAdapter.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var base = require('@web3auth/base');
var baseEvmAdapter = require('@web3auth/base-evm-adapter');

class InjectedEvmAdapter extends baseEvmAdapter.BaseEvmAdapter {
  constructor(options) {
    super(options);
    _defineProperty(this, "adapterNamespace", base.ADAPTER_NAMESPACES.EIP155);
    _defineProperty(this, "currentChainNamespace", base.CHAIN_NAMESPACES.EIP155);
    _defineProperty(this, "type", base.ADAPTER_CATEGORY.EXTERNAL);
    _defineProperty(this, "name", void 0);
    _defineProperty(this, "isInjected", true);
    _defineProperty(this, "status", base.ADAPTER_STATUS.NOT_READY);
    _defineProperty(this, "injectedProvider", null);
    this.name = options.name;
    this.injectedProvider = options.provider;
  }
  get provider() {
    if (this.status !== base.ADAPTER_STATUS.NOT_READY && this.injectedProvider) {
      return this.injectedProvider;
    }
    return null;
  }
  set provider(_) {
    throw new Error("Not implemented");
  }
  async init(options = {}) {
    await super.init(options);
    super.checkInitializationRequirements();
    this.status = base.ADAPTER_STATUS.READY;
    this.emit(base.ADAPTER_EVENTS.READY, this.name);
    try {
      base.log.debug(`initializing ${this.name} injected adapter`);
      if (options.autoConnect) {
        this.rehydrated = true;
        await this.connect();
      }
    } catch (error) {
      this.emit(base.ADAPTER_EVENTS.ERRORED, error);
    }
  }
  async connect() {
    super.checkConnectionRequirements();
    if (!this.injectedProvider) throw base.WalletLoginError.connectionError("Injected provider is not available");
    if (!this.chainConfig) throw base.WalletLoginError.connectionError("Chain config is not available");
    this.status = base.ADAPTER_STATUS.CONNECTING;
    this.emit(base.ADAPTER_EVENTS.CONNECTING, {
      adapter: this.name
    });
    try {
      await this.injectedProvider.request({
        method: "eth_requestAccounts"
      });
      // switch chain if not connected to the right chain
      if (this.injectedProvider.chainId !== this.chainConfig.chainId) {
        try {
          await this.switchChain(this.chainConfig, true);
        } catch (error) {
          await this.addChain(this.chainConfig, true);
          await this.switchChain(this.chainConfig, true);
        }
      }
      this.status = base.ADAPTER_STATUS.CONNECTED;
      const accountDisconnectHandler = accounts => {
        if (accounts.length === 0) {
          var _this$injectedProvide;
          this.disconnect();
          if ((_this$injectedProvide = this.injectedProvider) !== null && _this$injectedProvide !== void 0 && _this$injectedProvide.removeListener) this.injectedProvider.removeListener("accountsChanged", accountDisconnectHandler);
        }
      };
      this.injectedProvider.on("accountsChanged", accountDisconnectHandler);
      this.emit(base.ADAPTER_EVENTS.CONNECTED, {
        adapter: this.name,
        reconnected: this.rehydrated,
        provider: this.injectedProvider
      });
      return this.injectedProvider;
    } catch (error) {
      // ready again to be connected
      this.status = base.ADAPTER_STATUS.READY;
      this.rehydrated = false;
      this.emit(base.ADAPTER_EVENTS.ERRORED, error);
      if (error instanceof base.Web3AuthError) throw error;
      throw base.WalletLoginError.connectionError(`Failed to login with ${this.name} injected wallet`);
    }
  }
  async disconnect(options = {
    cleanup: false
  }) {
    if (!this.injectedProvider) throw base.WalletLoginError.connectionError("Injected provider is not available");
    await super.disconnectSession();
    if (typeof this.injectedProvider.removeAllListeners !== "undefined") this.injectedProvider.removeAllListeners();
    try {
      await this.injectedProvider.request({
        method: "wallet_revokePermissions",
        params: [{
          eth_accounts: {}
        }]
      });
    } catch (error) {}
    if (options.cleanup) {
      this.status = base.ADAPTER_STATUS.NOT_READY;
      this.injectedProvider = null;
    } else {
      // ready to be connected again
      this.status = base.ADAPTER_STATUS.READY;
    }
    await super.disconnect();
  }
  async getUserInfo() {
    if (this.status !== base.ADAPTER_STATUS.CONNECTED) throw base.WalletLoginError.notConnectedError("Not connected with wallet, Please login/connect first");
    return {};
  }
  async addChain(chainConfig, init = false) {
    if (!this.injectedProvider) throw base.WalletLoginError.connectionError("Injected provider is not available");
    super.checkAddChainRequirements(chainConfig, init);
    await this.injectedProvider.request({
      method: "wallet_addEthereumChain",
      params: [{
        chainId: chainConfig.chainId,
        chainName: chainConfig.displayName,
        rpcUrls: [chainConfig.rpcTarget],
        blockExplorerUrls: [chainConfig.blockExplorerUrl],
        nativeCurrency: {
          name: chainConfig.tickerName,
          symbol: chainConfig.ticker,
          decimals: chainConfig.decimals || 18
        },
        iconUrls: [chainConfig.logo]
      }]
    });
    this.addChainConfig(chainConfig);
  }
  async switchChain(params, init = false) {
    if (!this.injectedProvider) throw base.WalletLoginError.connectionError("Injected provider is not available");
    super.checkSwitchChainRequirements(params, init);
    await this.injectedProvider.request({
      method: "wallet_switchEthereumChain",
      params: [{
        chainId: params.chainId
      }]
    });
    this.setAdapterSettings({
      chainConfig: this.getChainConfig(params.chainId)
    });
  }
  async enableMFA() {
    throw new Error("Method Not implemented");
  }
  async manageMFA() {
    throw new Error("Method Not implemented");
  }
}

exports.InjectedEvmAdapter = InjectedEvmAdapter;
