// @web3auth/ui/dist/lib.cjs/packages/ui/src/components/SelfCustodyViaWeb3Auth.js
'use strict';

var reactI18next = require('react-i18next');
var localeImport = require('../localeImport.js');
var jsxRuntime = require('react/jsx-runtime');

function SelfCustodyViaWeb3Auth() {
  const [t] = reactI18next.useTranslation(undefined, {
    i18n: localeImport
  });
  return /*#__PURE__*/jsxRuntime.jsxs("div", {
    className: "w3a--flex w3a--items-center w3a--gap-2 w3a--justify-center",
    children: [/*#__PURE__*/jsxRuntime.jsx("div", {
      className: "w3a--text-xs w3a--text-app-gray-300 dark:w3a--text-app-gray-500",
      children: t("modal.footer.message-new")
    }), /*#__PURE__*/jsxRuntime.jsx("img", {
      height: "16",
      src: "https://images.web3auth.io/web3auth-footer-logo-light.svg",
      alt: "Web3Auth Logo Light",
      className: "w3a--h-4 w3a--block dark:w3a--hidden"
    }), /*#__PURE__*/jsxRuntime.jsx("img", {
      height: "16",
      src: "https://images.web3auth.io/web3auth-footer-logo-dark.svg",
      alt: "Web3Auth Logo Dark",
      className: "w3a--h-4 w3a--hidden dark:w3a--block"
    })]
  });
}

module.exports = SelfCustodyViaWeb3Auth;
