// @web3auth/ui/dist/lib.esm/packages/ui/src/components/Header.js
import { memo } from 'react';
import { useTranslation } from 'react-i18next';
import { DEFAULT_LOGO_DARK, DEFAULT_LOGO_LIGHT } from '../interfaces.js';
import i18nInstance from '../localeImport.js';
import Icon from './Icon.js';
import { jsxs, jsx } from 'react/jsx-runtime';

function Header(props) {
  const {
    onClose,
    appLogo,
    appName
  } = props;
  const [t] = useTranslation(undefined, {
    i18n: i18nInstance
  });
  const headerLogo = [DEFAULT_LOGO_DARK, DEFAULT_LOGO_LIGHT].includes(appLogo) ? "" : appLogo;
  const subtitle = t("modal.header-subtitle-name", {
    appName
  });
  return /*#__PURE__*/jsxs("div", {
    className: "w3a-modal__header",
    children: [/*#__PURE__*/jsx("div", {
      className: "w3a-header",
      children: /*#__PURE__*/jsxs("div", {
        children: [headerLogo && /*#__PURE__*/jsx("div", {
          className: "w3a-header__logo",
          children: /*#__PURE__*/jsx("img", {
            src: headerLogo,
            alt: ""
          })
        }), /*#__PURE__*/jsx("div", {
          className: "w3a-header__title",
          children: t("modal.header-title")
        }), /*#__PURE__*/jsxs("div", {
          className: "w3a-header__subtitle",
          children: [subtitle, /*#__PURE__*/jsxs("div", {
            className: "w3a--relative w3a--flex w3a--flex-col w3a--items-center w3a--group w3a--cursor-pointer",
            children: [/*#__PURE__*/jsx(Icon, {
              iconName: "information-circle-light",
              darkIconName: "information-circle"
            }), /*#__PURE__*/jsxs("div", {
              className: "w3a--absolute w3a--top-4 w3a--z-20 w3a--flex-col w3a--items-center w3a--hidden w3a--mb-5 group-hover:w3a--flex",
              children: [/*#__PURE__*/jsx("div", {
                className: "w3a--w-3 w3a--h-3 w3a--ml-[3px] -w3a--mb-2 w3a--rotate-45 w3a--bg-app-gray-50 dark:w3a--bg-app-gray-600"
              }), /*#__PURE__*/jsxs("div", {
                className: `w3a--relative w3a--p-4 w3a--w-[270px] w3a--translate-x-[-16px] w3a--text-xs w3a--leading-none w3a--text-app-white w3a--rounded-md w3a--bg-app-gray-50 dark:w3a--bg-app-gray-600 w3a--shadow-lg ${subtitle.length > 34 ? "-w3a--ml-[100px]" : ""}`,
                children: [/*#__PURE__*/jsx("div", {
                  className: "w3a--text-xs w3a--font-medium w3a--mb-1 w3a--text-app-gray-900 dark:w3a--text-app-white",
                  children: t("modal.header-tooltip-title")
                }), /*#__PURE__*/jsx("div", {
                  className: "w3a--text-xs w3a--text-app-gray-400",
                  children: t("modal.header-tooltip-desc")
                })]
              })]
            })]
          })]
        })]
      })
    }), /*#__PURE__*/jsx("button", {
      type: "button",
      onClick: onClose,
      className: "w3a-header__button w3ajs-close-btn ",
      children: /*#__PURE__*/jsx(Icon, {
        iconName: "x-light",
        darkIconName: "x-dark"
      })
    })]
  });
}
const memoizedHeader = /*#__PURE__*/memo(Header, (prevProps, nextProps) => {
  if (prevProps.appLogo !== nextProps.appLogo) {
    return true;
  }
  return false;
});
memoizedHeader.displayName = "Header";

export { memoizedHeader as default };
