// @web3auth/ui/dist/lib.cjs/packages/ui/src/index.js
'use strict';

var config = require('./config.js');
var interfaces = require('./interfaces.js');
var loginModal = require('./loginModal.js');
var utils = require('./utils.js');



exports.AUTH_PROVIDERS = config.AUTH_PROVIDERS;
exports.AUTH_PROVIDERS_NAMES = config.AUTH_PROVIDERS_NAMES;
exports.capitalizeFirstLetter = config.capitalizeFirstLetter;
exports.DEFAULT_LOGO_DARK = interfaces.DEFAULT_LOGO_DARK;
exports.DEFAULT_LOGO_LIGHT = interfaces.DEFAULT_LOGO_LIGHT;
exports.LOGIN_MODAL_EVENTS = interfaces.LOGIN_MODAL_EVENTS;
exports.MODAL_STATUS = interfaces.MODAL_STATUS;
exports.WALLET_CONNECT_LOGO = interfaces.WALLET_CONNECT_LOGO;
exports.LoginModal = loginModal.LoginModal;
exports.getAdapterSocialLogins = utils.getAdapterSocialLogins;
exports.getNetworkIconId = utils.getNetworkIconId;
exports.getUserCountry = utils.getUserCountry;
exports.getUserLanguage = utils.getUserLanguage;
exports.passwordlessBackendUrl = utils.passwordlessBackendUrl;
exports.validateImageUrl = utils.validateImageUrl;
exports.validatePhoneNumber = utils.validatePhoneNumber;
