// @web3auth/modal/dist/lib.cjs/config.js
'use strict';

var base = require('@web3auth/base');

const defaultSolanaDappModalConfig = {
  chainNamespace: base.CHAIN_NAMESPACES.SOLANA,
  adapters: {
    [base.SOLANA_ADAPTERS.TORUS_SOLANA]: {
      label: "Torus Wallet",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    },
    [base.SOLANA_ADAPTERS.AUTH]: {
      label: "Auth",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    }
  }
};
const defaultEvmDappModalConfig = {
  chainNamespace: base.CHAIN_NAMESPACES.EIP155,
  adapters: {
    [base.EVM_ADAPTERS.TORUS_EVM]: {
      label: "Torus Wallet",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    },
    [base.EVM_ADAPTERS.AUTH]: {
      label: "Auth",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    },
    [base.EVM_ADAPTERS.WALLET_CONNECT_V2]: {
      label: "Wallet Connect",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    }
  }
};
const defaultSolanaWalletModalConfig = {
  chainNamespace: base.CHAIN_NAMESPACES.SOLANA,
  adapters: {
    [base.SOLANA_ADAPTERS.AUTH]: {
      label: "Auth",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    }
  }
};
const defaultEvmWalletModalConfig = {
  chainNamespace: base.CHAIN_NAMESPACES.EIP155,
  adapters: {
    [base.EVM_ADAPTERS.AUTH]: {
      label: "Auth",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    }
  }
};
const defaultOtherModalConfig = {
  chainNamespace: base.CHAIN_NAMESPACES.OTHER,
  adapters: {
    [base.EVM_ADAPTERS.AUTH]: {
      label: "Auth",
      showOnModal: true,
      showOnMobile: true,
      showOnDesktop: true
    }
  }
};
const walletRegistryUrl = "https://assets.web3auth.io/v1/wallet-registry.json";

exports.defaultEvmDappModalConfig = defaultEvmDappModalConfig;
exports.defaultEvmWalletModalConfig = defaultEvmWalletModalConfig;
exports.defaultOtherModalConfig = defaultOtherModalConfig;
exports.defaultSolanaDappModalConfig = defaultSolanaDappModalConfig;
exports.defaultSolanaWalletModalConfig = defaultSolanaWalletModalConfig;
exports.walletRegistryUrl = walletRegistryUrl;
