// @web3auth/default-solana-adapter/dist/lib.cjs/walletStandardAdapter.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var features = require('@wallet-standard/features');
var base = require('@web3auth/base');
var baseSolanaAdapter = require('@web3auth/base-solana-adapter');
var solanaProvider = require('@web3auth/solana-provider');
var utils = require('./utils.js');
var walletStandardHandler = require('./walletStandardHandler.js');

class WalletStandardAdapter extends baseSolanaAdapter.BaseSolanaAdapter {
  constructor(options) {
    super(options);
    _defineProperty(this, "name", void 0);
    _defineProperty(this, "adapterNamespace", base.ADAPTER_NAMESPACES.SOLANA);
    _defineProperty(this, "currentChainNamespace", base.CHAIN_NAMESPACES.SOLANA);
    _defineProperty(this, "type", base.ADAPTER_CATEGORY.EXTERNAL);
    _defineProperty(this, "isInjected", true);
    _defineProperty(this, "status", base.ADAPTER_STATUS.NOT_READY);
    _defineProperty(this, "wallet", null);
    _defineProperty(this, "injectedProvider", null);
    this.name = options.name;
    // in VueJS, for some wallets e.g. Gate, Solflare, when connecting it throws error "attempted to get private field on non-instance"
    // it seems that Vue create a Proxy object for the wallet object which causes the issue
    // ref: https://stackoverflow.com/questions/64917686/vue-array-converted-to-proxy-object
    this.wallet = ["gate", "solflare"].includes(this.name) ? Object.freeze(options.wallet) : options.wallet;
  }
  get provider() {
    if (this.status !== base.ADAPTER_STATUS.NOT_READY && this.injectedProvider) {
      return this.injectedProvider;
    }
    return null;
  }
  get isWalletConnected() {
    return !!(this.status === base.ADAPTER_STATUS.CONNECTED && this.wallet.accounts.length > 0);
  }
  async init(options = {}) {
    await super.init(options);
    super.checkInitializationRequirements();
    this.injectedProvider = new solanaProvider.WalletStandardProvider({
      config: {
        chainConfig: this.chainConfig
      }
    });
    const providerHandler = new walletStandardHandler.WalletStandardProviderHandler({
      wallet: this.wallet,
      getCurrentChain: () => utils.getSolanaChainByChainConfig(this.chainConfig)
    });
    this.injectedProvider.setupProvider(providerHandler);
    this.status = base.ADAPTER_STATUS.READY;
    this.emit(base.ADAPTER_EVENTS.READY, this.name);
    try {
      base.log.debug("initializing solana injected adapter");
      if (options.autoConnect) {
        this.rehydrated = true;
        await this.connect();
      }
    } catch (error) {
      base.log.error("Failed to connect with cached solana injected provider", error);
      this.emit(base.ADAPTER_EVENTS.ERRORED, error);
    }
  }
  async connect() {
    try {
      super.checkConnectionRequirements();
      this.status = base.ADAPTER_STATUS.CONNECTING;
      this.emit(base.ADAPTER_EVENTS.CONNECTING, {
        adapter: this.name
      });
      const chainName = utils.getSolanaChainByChainConfig(this.chainConfig);
      if (!this.wallet.chains.find(chain => chain === chainName)) throw base.WalletLoginError.connectionError(`Chain ${chainName} not supported. Supported chains are ${this.wallet.chains.join(", ")}`);
      if (!this.isWalletConnected) {
        await this.wallet.features[features.StandardConnect].connect();
      }
      if (this.wallet.accounts.length === 0) throw base.WalletLoginError.connectionError();
      this.status = base.ADAPTER_STATUS.CONNECTED;
      this.emit(base.ADAPTER_EVENTS.CONNECTED, {
        adapter: this.name,
        reconnected: this.rehydrated,
        provider: this.provider
      });
      return this.provider;
    } catch (error) {
      // ready again to be connected
      this.status = base.ADAPTER_STATUS.READY;
      this.rehydrated = false;
      this.emit(base.ADAPTER_EVENTS.ERRORED, error);
      throw error;
    }
  }
  async disconnect(options = {
    cleanup: false
  }) {
    await super.disconnectSession();
    try {
      var _this$wallet$features;
      await ((_this$wallet$features = this.wallet.features[features.StandardDisconnect]) === null || _this$wallet$features === void 0 ? void 0 : _this$wallet$features.disconnect());
      if (options.cleanup) {
        this.status = base.ADAPTER_STATUS.NOT_READY;
        this.injectedProvider = null;
      } else {
        this.status = base.ADAPTER_STATUS.READY;
      }
      await super.disconnect();
    } catch (error) {
      this.emit(base.ADAPTER_EVENTS.ERRORED, base.WalletLoginError.disconnectionError(error === null || error === void 0 ? void 0 : error.message));
    }
  }
  async getUserInfo() {
    if (!this.isWalletConnected) throw base.WalletLoginError.notConnectedError("Not connected with wallet, Please login/connect first");
    return {};
  }
  async addChain(chainConfig, init = false) {
    var _this$injectedProvide;
    super.checkAddChainRequirements(chainConfig, init);
    (_this$injectedProvide = this.injectedProvider) === null || _this$injectedProvide === void 0 || _this$injectedProvide.addChain(chainConfig);
    this.addChainConfig(chainConfig);
  }
  async switchChain(params, init = false) {
    var _this$injectedProvide2;
    super.checkSwitchChainRequirements(params, init);
    await ((_this$injectedProvide2 = this.injectedProvider) === null || _this$injectedProvide2 === void 0 ? void 0 : _this$injectedProvide2.switchChain(params));
    this.setAdapterSettings({
      chainConfig: this.getChainConfig(params.chainId)
    });
  }
  async enableMFA() {
    throw new Error("Method not implemented.");
  }
  async manageMFA() {
    throw new Error("Method Not implemented");
  }
}

exports.WalletStandardAdapter = WalletStandardAdapter;
